nix/tests.nix

```
@@ -217,291 +179,533 @@ let
+        cloud_hypervisor.succeed("mount /dev/sda1 /mnt")
+        cloud_hypervisor.succeed("echo 123TEST123 > /mnt/file.txt")
+        cloud_hypervisor.succeed("umount /mnt")
+        cloud_hypervisor.succeed("mount /dev/sda1 /mnt")
+        out = cloud_hypervisor.succeed("cat /mnt/file.txt")
+        search("123TEST123", out)
       '';
-
-      nodes.machine = _:
-        {
-          imports = [ testMachineConfig.basicMachineConfig testMachineConfig.systemdServices ];
-
-          services.udev.extraRules = ''
-            ACTION=="add|change", SUBSYSTEM=="usb", ATTRS{product}=="QEMU USB Keyboard", ATTRS{idVendor}=="${hid-vendorId}", ATTRS{idProduct}=="${hid-productId}", MODE="0660", GROUP="usbaccess", SYMLINK+="bus/usb/testdevice"
-          '';
-
-          virtualisation = {
-            cores = 2;
-            memorySize = 4096;
-            # QEMU QMP send-key commands will be sent through default non-usb (vfio) devices if not explicitly disabled
-            qemu.virtioKeyboard = false;
-            qemu.options = [
-              # A virtual USB UHCI controller in the host ...
-              "-device qemu-xhci,id=xhci,addr=10"
-              # ... with an attached usb keyboard.
-              "-device usb-kbd,bus=xhci.0"
-
-              # Enable QEMU QMP interactions.
-              "-chardev socket,id=qmp,path=/tmp/qmp.sock,server=on,wait=off"
-              "-mon chardev=qmp,mode=control,pretty=on"
-            ];
-          };
-        };
     };
+    ```
+
+  */
+  mkUsbvfiodTest = args: mkUsbvfiodTestInsecure (sanityCheckArgs args);
 
-  multiple-blockdevices =
-    pkgs.testers.runNixOSTest {
-      name = "usbvfiod with multiple blockdevices";
+  # See mkUsbvfiodTest (this runs without any arg checks).
+  mkUsbvfiodTestInsecure =
+    let
+      ehciProductName = "EHCI Host Controller";
+      xhciProductName = "xHCI Host Controller";
+    in
+    args: pkgs.testers.runNixOSTest {
+      inherit (args) name;
 
       inherit globalTimeout passthru;
 
       nodes.machine = _: {
-        imports = [ testMachineConfig.basicMachineConfig ];
+        imports = [ basicMachineConfig ];
+
+        # Create a udev rule for every device listed that enables it.
+        services.udev.extraRules =
+          lib.concatStrings (
+            builtins.map
+              (element:
+                if element.udevRule.enable then
+                  let
+                    controller =
+                      if (element.usbVersion == "2") then "${ehciProductName}"
+                      else if (element.usbVersion == "3") then "${xhciProductName}"
+                      else builtins.abort "only USB versions 2 using an ehci and 3 using a xhci controller are available";
+                    usbPort = builtins.toString element.usbPort;
+                  in
+                  if (usbPort == "" || element.udevRule.symlink == "")
+                  then abort "A udev rules requires to set a usbPort and a symlink string"
```

The aborts seem to contradict the statement about not doing argument checks here. Wouldn't this cause a failure elsewhere already? A plain attribute lookup may suffice here at least for the controller name.


```suggestion
                    controller = { "2" = ehciProductName; "3" = xhciProductName; }.${element.usbVersion};
                    usbPort = builtins.toString element.usbPort;
                  in
                  if (usbPort == "" || element.udevRule.symlink == "")
                  then abort "A udev rules requires to set a usbPort and a symlink string"
```